File: fiddling/static/systems/collision_system.js
import { System } from './systems.js'
import { collider_components } from '../components/collider_components.js'
import { garbage_filter } from '../utils.js'

class CollisionSystem extends System {
    constructor(){
	super()
	this.components_dict = collider_components
    }
    update(){
	// clean up
	for(var collider of this.components){
	    collider.colliding_with = []
	    collider.is_colliding = false
	}
	this.garbage_collect()
	
	for(var i = 0; i<this.components.length; i++){
	    for(var j = i+1; j<this.components.length; j++){
		var collider1 = this.components[i]
		var collider2 = this.components[j]
		if(collider1.collides_with(collider2)){
		    collider1.is_colliding = true
		    collider2.is_colliding = true
		    collider1.colliding_with.push(collider2)
		    collider2.colliding_with.push(collider1)
		}
	    }
	}
    }
}

export var collision_system = new CollisionSystem()

File: fiddling/server/systems/collision_system.py
from .systems import System
from server.components.collider_components import collider_components
from server.cutils import collisions
from typing import NamedTuple

class Collider(NamedTuple):
    typ: str
    x: float
    y: float
    rot: float
    size_x: float
    size_y: float
    r: float
    l: float


class CollisionSystem(System):
    def __init__(self):
        super().__init__()
        self.components_dict = collider_components

    def update(self, world, dt):
        for collider in self.components:
            collider.colliding_with = []
            collider.is_colliding = False
            collider.penetrations = []
        self.garbage_collect()

        colliders = []
        for component in self.components:
            absolute_pos = component.entity.position.get_absolute_pos()
            colliders.append(
                Collider(
                    typ=component.type.encode(),
                    x=absolute_pos.x,
                    y=absolute_pos.y,
                    rot=absolute_pos.rot,
                    size_x=absolute_pos.size_x,
                    size_y=absolute_pos.size_y,
                    r=component.radius if hasattr(component, "radius") else 0,
                    l=component.length if hasattr(component, "length") else 0
                )
            )

        collision_responses = collisions.solve(colliders)

        for collider, col_resp in zip(self.components, collision_responses):
            collider.is_colliding = bool(col_resp[0])
            for i, p in zip(col_resp[1], col_resp[2]):
                collider.colliding_with.append(self.components[i])
                collider.penetrations.append(p)


collision_system = CollisionSystem()
